### PR TITLE
Enable typechecking for _inductor/virtualized.py

### PR DESCRIPTION
**Pull request by int3**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109176
* #109173
* __->__ #108916

Also add a few more type annotations to utils.py (some of its functions
are called from virtualized.py)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/_inductor/dependencies.py [hunk 1]
```diff
@@ -189,7 +189,7 @@ def reads_and_writes(self):
         return itertools.chain(self.reads, self.writes)
 
 
-class _RecordLoadStoreInner(V.MockHandler):
+class _RecordLoadStoreInner(V.MockHandler):  # type: ignore[name-defined]
```
Comment (int3) [inline torch/_inductor/dependencies.py:192]: not clear to me why mypy thinks V.MockHandler is undefined in this context -- yet it's fine with e.g. a `foo = V.MockHandler` statement

**Review by eellison — APPROVED** [review]
sweet!

**Pull request closed without merging**